packages/proxy/lib/http/response-middleware.ts

```diff
@@ -231,16 +231,26 @@ const PatchExpressSetHeader: ResponseMiddleware = function () {
 
 const MaybeDelayForMultiDomain: ResponseMiddleware = function () {
   const isCrossDomain = !reqMatchesOriginPolicy(this.req, this.getRemoteState())
+  const isPreviousOrigin = this.getOriginStack().includes(cors.getOriginPolicy(this.req.proxiedUrl))
   const isHTML = resContentTypeIs(this.incomingRes, 'text/html')
   const isRenderedHTML = reqWillRenderHtml(this.req)
   const isAUTFrame = this.req.isAUTFrame
 
-  if (this.config.experimentalMultiDomain && isCrossDomain && isAUTFrame && (isHTML || isRenderedHTML)) {
-    this.debug('is cross-domain, delay until domain:ready event')
+  // delay the response if this is a cross-origin (and not returning to a previous origin) html request from the AUT iframe
+  if (this.config.experimentalMultiDomain && isCrossDomain && !isPreviousOrigin && isAUTFrame && (isHTML || isRenderedHTML)) {
+    this.debug('is cross-domain, delay until ready:for:domain(:failed) event')
+
+    this.serverBus.once('ready:for:domain:failed', () => {
+      this.debug('ready for domain failed, let it go')
+
+      this.next()
+    })
 
     this.serverBus.once('ready:for:domain', () => {
       this.debug('ready for domain, let it go')
 
+      this.res.wantsInjection = 'fullMultiDomain'
+
       this.next()
     })
```

Comment:
Since there are 2 events, when one gets called, the other will still have a listener attached and we'll end up with stacked up listeners on subsequent calls, which could lead to unexpected behavior. I think there are 2 options for fixing it:

- Remove the event listener for the other when one of them gets called
- Have only one event, that takes an argument to determine whether to inject or not

I personally prefer the second option.

Reply:
Went with the 2nd option. Fixed here: [145ab0e](https://github.com/cypress-io/cypress/pull/20752/commits/145ab0ecf31f25e2ee69ba400fa5c662156019f7)